### PR TITLE
Add self segmentation stable diffusion

### PR DESCRIPTION
**Pull request by orpatashnik**
@sayakpaul As to what we have been talking yesterday, I am opening a new pull request

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hey @orpatashnik,

Thanks a lot for the PR :-) 

Generally ok to add for me, but our pipelines are exploding a bit. I wonder whether we should move the more researchy pipelines better to the community folder. Do we expect this pipeline to be used in production?

**Comment by orpatashnik** [comment]
I think the usage is difficult to predict. So whatever you think is best :)

**Comment by sayakpaul** [comment]
> I wonder whether we should move the more researchy pipelines better to the community folder.

Could make sense for this pipeline IMO. If we start seeing usage, we could definitely graduate that in a regular pipeline. This way, the maintenance burden is also a bit reduced and you get to have a lot of flexibility in terms of code, @orpatashnik. 

But since this PR is already in good shape (except for a few minor things), I would not mind accommodating it, @patrickvonplaten. But going forward, I think we'd need to be a bit careful.

File: docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+between `kmeans` and `spectral`. The default is `spectral`.
+- `num_segments` - The number of clusters to use for the clustering algorithm. The default is 6. Since the method merges
+clusters, it is recommended to use a higher number of clusters than the actual number of desired segments.
+- `seg_labels_indices_in_prompt` (optional) - The indices of the tokens in the prompt that correspond to the labels of the segments.
+To label the segments, we consider the cross-attention maps in these indices and for each segment, we find the token
+whose cross-attention map has the highest intersection with the segment. This argument accepts a list of `int` as input.
+If not provided, we automatically extract indices of nouns from the prompt and use them.
+
+### Outputs
+In addition to the outputted generated image, this pipeline outputs the following:
```
Comment (mishig25) [inline docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx:75]:
```suggestion
### Outputs

In addition to the outputted generated image, this pipeline outputs the following:
```

File: docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+
+## Available Pipelines:
+
+| Pipeline | Tasks 
+|---|---
+| [SelfSegmentationStableDiffusionPipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/self_segmentation_stable_diffusion_pipeline.py) | *Text-to-Image Generation, Segmentation* | - |
+
+
+## Usage
+The usage of this pipeline is very similar to the vanilla Stable Diffusion pipeline. We describe the differences below.
```
Comment (mishig25) [inline docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx:59]:
```suggestion
## Usage

The usage of this pipeline is very similar to the vanilla Stable Diffusion pipeline. We describe the differences below.
```

File: docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+
+### Inputs
+- `segmentation_resolution` - This is the resolution of the self-attention maps used for the clustering. The default is
```
Comment (mishig25) [inline docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx:62]:
```suggestion
### Inputs

- `segmentation_resolution` - This is the resolution of the self-attention maps used for the clustering. The default is
```

File: docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+## FAQ
+
+### How well it performs compared to dedicated segmentation networks?
```
Comment (mishig25) [inline docs/source/en/api/pipelines/self_segmentation_stable_diffusion.mdx:35]:
```suggestion
### How well it performs compared to dedicated segmentation networks?

```

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 5]
```diff
@@ -0,0 +1,1012 @@
...
+from ...pipelines.pipeline_utils import DiffusionPipeline
+from ...pipelines.stable_diffusion import StableDiffusionSafetyChecker
+from ...schedulers import KarrasDiffusionSchedulers
+from . import SelfSegmentationStableDiffusionPipelineOutput
+
+
+if is_sklearn_available():
+    from sklearn.cluster import KMeans, SpectralClustering
+if is_nltk_available():
+    import nltk
```
Comment (patrickvonplaten) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:49]: what happens if one tries to run the pipeline with `nltk`? Does it error out? In this case let's maybe make sure to only import the pipeline if `nltk` is installed

File: docs/source/en/api/pipelines/overview.mdx [hunk 6]
```diff
@@ -90,6 +90,7 @@ available a colab notebook to directly try them out.
 | [versatile_diffusion](./versatile_diffusion) | [**Versatile Diffusion: Text, Images and Variations All in One Diffusion Model**](https://arxiv.org/abs/2211.08332) | Dual Image and Text Guided Generation |
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/overview.mdx:90]: Let's change the format to `md`.

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py (hunk 5 above)
Comment (sayakpaul) [reply src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:49]: Agreed. For that to happen I think the check needs to go to `__init__.py` of the pipeline, no?

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 7]
```diff
@@ -0,0 +1,1012 @@
...
+        self.vae.enable_tiling()
+
+    def disable_vae_tiling(self):
+        r"""
+        Disable tiled VAE decoding. If `enable_vae_tiling` was previously invoked, this method will go back to
+        computing decoding in one step.
+        """
+        self.vae.disable_tiling()
+
+    def enable_sequential_cpu_offload(self, gpu_id=0):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:343]: We can remove this method as `DiffuserPipeline` now has this method enabled.

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 8]
```diff
@@ -0,0 +1,1012 @@
...
+            _, hook = cpu_offload_with_hook(cpu_offloaded_model, device, prev_module_hook=hook)
+
+        if self.safety_checker is not None:
+            _, hook = cpu_offload_with_hook(self.safety_checker, device, prev_module_hook=hook)
+
+        # We'll offload the last model manually.
+        self.final_offload_hook = hook
+
+    @property
+    def _execution_device(self):
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:397]: Same. We can remove it.

File: src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py [hunk 9]
```diff
@@ -0,0 +1,1012 @@
...
+        callback_steps: int = 1,
+        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
+        guidance_rescale: float = 0.0,
+        segmentation_resolution: int = 32,
+        segmentation_clustering_algorithm: str = "spectral",
+        num_segments: int = 6,
+        seg_labels_indices_in_prompt: Optional[List[int]] = None,
+    ):
+        r"""
+        Function invoked when calling the pipeline for generation.
```
Comment (sayakpaul) [inline src/diffusers/pipelines/self_segmentation_stable_diffusion/self_segmentation_stable_diffusion.py:779]: Let's maybe make it more descriptive?

File: src/diffusers/utils/dummy_torch_and_transformers_objects.py [hunk 10]
```diff
@@ -362,6 +362,21 @@ def from_pretrained(cls, *args, **kwargs):
...
+    def __init__(self, *args, **kwargs):
+        requires_backends(self, ["torch", "transformers"])
+
+    @classmethod
+    def from_config(cls, *args, **kwargs):
+        requires_backends(cls, ["torch", "transformers"])
+
+    @classmethod
+    def from_pretrained(cls, *args, **kwargs):
+        requires_backends(cls, ["torch", "transformers"])
```
Comment (sayakpaul) [inline src/diffusers/utils/dummy_torch_and_transformers_objects.py:377]: Hmm, there needs to be an NLTK entry for this as well as it's a separate backend.

File: tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+    def test_inference_batch_consistent(self, batch_sizes=[2, 4, 13]):
+        super().test_inference_batch_consistent(batch_sizes=batch_sizes)
+
+    def test_inference_batch_single_identical(self, batch_size=3, expected_max_diff=1e-2):
+        super().test_inference_batch_single_identical(batch_size=batch_size, expected_max_diff=expected_max_diff)
+
+
+@slow
+@require_torch_gpu
+class SelfSegmentationStableDiffusionSlowTests(unittest.TestCase):
```
Comment (sayakpaul) [inline tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py:199]: Fine to discard the SLOW tests for this pipeline.

File: tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        assert image.shape == (3, 128, 128, 3)
+        assert len(seg_maps) == len(seg_labels) == 3
+        assert len(merged_seg_maps) == len(merged_seg_labels) == 3
+
+    def test_float16_inference(self, expected_max_diff=1e-2):
+        pass
+
+    def test_karras_schedulers_shape(self):
+        pass
+
```
Comment (sayakpaul) [inline tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py:189]: Do we know why this does not work?

Comment (orpatashnik) [reply tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py:186]: My pipeline relies on an attention processor that saves attention maps. After all the denoising steps, I use these maps to segment. For some reason, in these two tests the attention maps that I save disappear after the unet returns its output (at each denoising step). For the float16, I tested the pipeline and it does work.

File: docs/source/en/api/pipelines/overview.md [hunk 13]
```diff
@@ -12,25 +12,108 @@ specific language governing permissions and limitations under the License.
 
 # Pipelines
 
-Pipelines provide a simple way to run state-of-the-art diffusion models in inference by bundling all of the necessary components (multiple independently-trained models, schedulers, and processors) into a single end-to-end class. Pipelines are flexible and they can be adapted to use different scheduler or even model components.
+Pipelines provide a simple way to run state-of-the-art diffusion models in inference.
+Most diffusion systems consist of multiple independently-trained models and highly adaptable scheduler 
+components - all of which are needed to have a functioning end-to-end diffusion system.
```
Comment (sayakpaul) [inline docs/source/en/api/pipelines/overview.md:17]: This file should not be affected by this PR.

File: tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py (hunk 12 above)
Comment (sayakpaul) [reply tests/pipelines/self_segmentation_stable_diffusion/test_self_segementation_stable_diffusion.py:186]: I think it's fine then to disable them for now.

File: docs/source/en/api/pipelines/overview.md (hunk 13 above)
Comment (orpatashnik) [reply docs/source/en/api/pipelines/overview.md]: it does not affect it now :)

**Comment by sayakpaul** [comment]
@yiyixuxu could you do a review here when you get a moment? 

I think we need to figure out how we handle the pipeline import when `nltk` is not installed properly. We would probably need to do something like: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion_xl/__init__.py.

**Comment by orpatashnik** [comment]
@yiyixuxu Hi ☺️ do you think you could have time for reviewing soon?

**Review by DN6 — COMMENTED** [review]
LGTM 👍🏽

**Comment by sayakpaul** [comment]
@orpatashnik since the pipeline requires `nltk` to be fully functional, we need to ensure our soft dependency system is on point. You can refer to https://github.com/huggingface/diffusers/blob/b37dc3b3cdda401ee370a5e2a8d2e27a016078ac/src/diffusers/pipelines/stable_diffusion_xl/__init__.py to get a sense of how to ensure `is_nltk_available()` should be used here in the context. 

Once this change is incorporated I think we can proceed towards merging :)

**Pull request closed without merging**